packages/flutter_tools/lib/src/plugins.dart

```diff
@@ -26,41 +28,158 @@ class Plugin {
+    }
+    if (pluginYaml != null && pluginYaml['platforms'] != null) {
+      return Plugin._fromMultiPlatformYaml(name, path, pluginYaml);
+    } else {
+      return Plugin._fromLegacyYaml(name, path, pluginYaml); // ignore: deprecated_member_use_from_same_package
+    }
+  }
+
+  factory Plugin._fromMultiPlatformYaml(String name, String path, dynamic pluginYaml) {
+    assert (pluginYaml != null && pluginYaml['platforms'] != null,
+            'Invalid multi-platform plugin specification.');
+    final dynamic platformsYaml = pluginYaml['platforms'];
+
+    assert (_validateMultiPlatformYaml(platformsYaml).isEmpty,
+            'Invalid multi-platform plugin specification.');
+
+    final Map<String, PluginPlatform> platforms = <String, PluginPlatform>{};
+
+    if (platformsYaml[AndroidPlugin.kConfigKey] != null) {
+      platforms[AndroidPlugin.kConfigKey] =
+          AndroidPlugin.fromYaml(name, platformsYaml[AndroidPlugin.kConfigKey]);
+    }
+
+    if (platformsYaml[IOSPlugin.kConfigKey] != null) {
+      platforms[IOSPlugin.kConfigKey] =
+          IOSPlugin.fromYaml(name, platformsYaml[IOSPlugin.kConfigKey]);
+    }
+
+    if (platformsYaml[MacOSPlugin.kConfigKey] != null) {
+      platforms[MacOSPlugin.kConfigKey] =
+          MacOSPlugin.fromYaml(name, platformsYaml[MacOSPlugin.kConfigKey]);
+    }
+
+    return Plugin(
+      name: name,
+      path: path,
+      platforms: platforms,
+    );
+  }
+
+  @deprecated
+  factory Plugin._fromLegacyYaml(String name, String path, dynamic pluginYaml) {
+    final Map<String, PluginPlatform> platforms = <String, PluginPlatform>{};
+    final String pluginClass = pluginYaml['pluginClass'];
+    if (pluginYaml != null && pluginClass != null) {
+      final String androidPackage = pluginYaml['androidPackage'];
+      if (androidPackage != null) {
+        platforms[AndroidPlugin.kConfigKey] =
+            AndroidPlugin(
+              name: name,
+              package: pluginYaml['androidPackage'],
+              pluginClass: pluginClass,
+            );
+      }
+
+      final String iosPrefix = pluginYaml['iosPrefix'] ?? '';
+      platforms[IOSPlugin.kConfigKey] =
+          IOSPlugin(
+            name: name,
+            classPrefix: iosPrefix,
+            pluginClass: pluginClass,
+          );
     }
     return Plugin(
       name: name,
       path: path,
-      androidPackage: androidPackage,
-      iosPrefix: iosPrefix,
-      macosPrefix: macosPrefix,
-      pluginClass: pluginClass,
+      platforms: platforms,
     );
   }
 
+  static List<String> validatePluginYaml(YamlMap yaml) {
+    if (yaml.containsKey('platforms')) {
+      final int numKeys = yaml.keys.toSet().length;
+      if (numKeys != 1) {
+        return <String>[
+          'Invalid plugin specification. There must be only one key: "platforms", found multiple: ${yaml.keys.join(',')}'
+        ];
+      } else {
+        return _validateMultiPlatformYaml(yaml['platforms']);
```

Comment:
If `yaml['platforms']` is not a `YamlMap`, then this call will throw a runtime type error. Please add a check, and return an error instead.